blog/2021/01/snoopi_deep.md

```diff
@@ -0,0 +1,221 @@
+We'll see how to deal with these in later installments.
+
+You also see breaks in the flamegraph. During these periods, code generation and runtime create new objects and then call methods on those objects; if one of those calls requires a fresh entrance into inference, that triggers the creation of a new flame.
+Hence, the number of distinct flames (which is just equal to `length(tinf.children)`) gives you a rough indication of how frequently the chains of inference were broken.
+
+## Elementary analysis: `flatten` and `accumulate_by_source`
+
+As our last step for this post, let's extract the data as a list:
+
+```
+julia> flatten(tinf)
+10-element Vector{SnoopCompileCore.InferenceTiming}:
+ InferenceTiming: 0.000034/0.000034 on InferenceFrameInfo for getproperty(::Main.SnoopDemo.MyType{Int64}, x::Symbol)
+ InferenceTiming: 0.000045/0.000045 on InferenceFrameInfo for getproperty(::Main.SnoopDemo.MyType{Int64}, ::Symbol)
+ InferenceTiming: 0.000071/0.000071 on InferenceFrameInfo for Main.SnoopDemo.MyType(::Int64)
+ InferenceTiming: 0.000083/0.000161 on InferenceFrameInfo for Main.SnoopDemo.extract(::Main.SnoopDemo.MyType{Int64})
+ InferenceTiming: 0.000122/0.000453 on InferenceFrameInfo for Main.SnoopDemo.dostuff(::Main.SnoopDemo.MyType{Int64})
+ InferenceTiming: 0.000122/0.000254 on InferenceFrameInfo for Main.SnoopDemo.MyType(::Int64)
+ InferenceTiming: 0.000132/0.000132 on InferenceFrameInfo for Main.SnoopDemo.MyType{Int64}(::Int64)
+ InferenceTiming: 0.000170/0.000170 on InferenceFrameInfo for Main.SnoopDemo.domath(::Int64)
+ InferenceTiming: 0.000355/0.001133 on InferenceFrameInfo for Main.SnoopDemo.packintype(::Int64)
+ InferenceTiming: 0.009382/0.010515 on InferenceFrameInfo for Core.Compiler.Timings.ROOT()
+```
+
+By default, this orders the nodes by exclusive time, but `flatten(tinf; sortby=inclusive)` allows you to sort by inclusive time.
+Finally, `flatten(tinf; sortby=nothing)` returns the nodes in the order in which they were inferred.
+
+Sometimes, you might infer the same method for many different specific argument types, and sometimes you might want to get a sense of the aggregate cost:
+
+```
+julia> accumulate_by_source(flatten(tinf))
+8-element Vector{Tuple{Float64, Union{Method, Core.MethodInstance}}}:
+ (7.838100000000001e-5, getproperty(x, f::Symbol) in Base at Base.jl:33)
+ (8.2955e-5, extract(y::Main.SnoopDemo.MyType) in Main.SnoopDemo at REPL[1]:4)
+ (0.000121738, dostuff(y) in Main.SnoopDemo at REPL[1]:11)
+ (0.000132328, Main.SnoopDemo.MyType{T}(x) where T in Main.SnoopDemo at REPL[1]:2)
+ (0.000170205, domath(x) in Main.SnoopDemo at REPL[1]:6)
+ (0.000193107, Main.SnoopDemo.MyType(x::T) where T in Main.SnoopDemo at REPL[1]:2)
+ (0.000354527, packintype(x) in Main.SnoopDemo at REPL[1]:13)
+ (0.009381595, ROOT() in Core.Compiler.Timings at compiler/typeinfer.jl:75)
+```
+
+This shows the cost of each `Method`, summing across all specific `MethodInstance`s.
```

Comment:
We've found ourselves issuing queries like this, quite a bit:

```julia
julia> fs = [t for t in flattened if occursin("merge_sorted!", repr(t[2].mi.def.name));

julia> sum(first.(fs))  # what is the total time inferring all specializations of all methods of merge_sorted! 
341.62921501600005

julia> length(fs)  # how many specializations are there?
5968
```
Or using this to plot a distribution of the specialization times for this function:
<img width="480" alt="Screen Shot 2021-01-20 at 5 49 37 PM" src="https://user-images.githubusercontent.com/1582097/105250525-e94ee700-5b47-11eb-8927-dfefb674036c.png">

Reply:
I'm not sure if it's worth mentioning things like this since we don't have standard built-in functions for things like this, and also hopefully most people are better off than we are, but 🤷 i thought i'd mention it :)

Reply:
`pgdsgui` will get at such issues, I favor leaving it for later.